### PR TITLE
Enable GAPI VASOT in Python

### PR DESCRIPTION
**Pull request by AsyaPronina**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by dmatveev** [comment]
Various build issues in the last build.. Also failures detected

File: modules/gapi/misc/python/python_bridge.hpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include <opencv2/gapi/garg.hpp>
 #include <opencv2/gapi/gopaque.hpp>
 #include <opencv2/gapi/render/render_types.hpp> // Prim
+#include <opencv2/gapi/ot_tracking_status.hpp> // TrackingStatus
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/python_bridge.hpp:14]: Should it be just a part of `ot.hpp`?

**Review by TolyaTalamanov — COMMENTED** [review]
The changes looks good but invasive. I'm not sure that `OpaqueKind` was ever designed to contain such specific detail as `TrackingStatus`. 

Perhaps we can avoid exposing `TrackingStatus` into python and just use `int` instead. For improving readability we still can define `TrackingStatus` enum but on python side (`__init__.py`) and use it like that:

```
rois, labels, ids, status = cv.gapi.ot.track(img, rect, labels, delta)
if status == TrackingStatus.NEW.value:
    pass
```

In that case the `status` is `int` but you may compare it with `enum`:
```
class TrackingStatus(Enum):
    NEW = 1,
    TRACKED = 2,
    LOST = 3
```

**Comment by TolyaTalamanov** [comment]
> The changes looks good but invasive. I'm not sure that `OpaqueKind` was ever designed to contain such specific detail as `TrackingStatus`.
> 
> Perhaps we can avoid exposing `TrackingStatus` into python and just use `int` instead. For improving readability we still can define `TrackingStatus` enum but on python side (`__init__.py`) and use it like that:
> 
> ```
> rois, labels, ids, status = cv.gapi.ot.track(img, rect, labels, delta)
> if status == TrackingStatus.NEW.value:
>     pass
> ```
> 
> In that case the `status` is `int` but you may compare it with `enum`:
> 
> ```
> class TrackingStatus(Enum):
>     NEW = 1,
>     TRACKED = 2,
>     LOST = 3
> ```

Need to discuss this, since it's not the better solution and in some extent even worse :)

File: modules/gapi/misc/python/package/gapi/__init__.py [hunk 2]
```diff
@@ -299,3 +332,10 @@ def kernel_with_params(cls):
 
 
 cv.gapi.wip.GStreamerPipeline = cv.gapi_wip_gst_GStreamerPipeline
+
+class TrackingStatus:
```
Comment (TolyaTalamanov) [inline modules/gapi/misc/python/package/gapi/__init__.py:336]: (Optional) 

What type is used for `NEW`, `LOST`, `TRACKED`? I assume `int`, will there be just `0`,`1`,`2` when we print it into console?
Perhaps we can leverage python `enum` module to improve that.

File: modules/gapi/src/backends/common/serialization.cpp [hunk 3]
```diff
@@ -366,6 +366,17 @@ IIStream& operator>> (IIStream& is,       cv::gapi::wip::draw::Line &l) {
     return is >> l.color >> l.lt >> l.pt1 >> l.pt2 >> l.shift >> l.thick;
 }
 
+IIStream& operator>> (IIStream& is, cv::gapi::ot::TrackingStatus& status) {
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/common/serialization.cpp:369]: Why not just `static_cast<int>`?

**Review by TolyaTalamanov — COMMENTED** [review]
@AsyaPronina Could you add a couple of python test cases, please?

File: modules/gapi/include/opencv2/gapi/gcommon.hpp [hunk 4]
```diff
@@ -41,22 +42,23 @@ namespace detail
...
+        CV_STRING,             // std::string user G-API data
+        CV_POINT,              // cv::Point user G-API data
+        CV_POINT2F,            // cv::Point2f user G-API data
+        CV_POINT3F,            // cv::Point3f user G-API data
+        CV_SIZE,               // cv::Size user G-API data
+        CV_RECT,               // cv::Rect user G-API data
+        CV_SCALAR,             // cv::Scalar user G-API data
+        CV_MAT,                // cv::Mat user G-API data
+        CV_DRAW_PRIM,          // cv::gapi::wip::draw::Prim user G-API data
+        CV_OT_TRACKING_STATUS  // cv::gapi::ot::TrackingStatus user G-API data
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gcommon.hpp:61]: I doubt if we need to know about this type at the Opaque level.

File: modules/gapi/misc/python/python_bridge.hpp (hunk 1 above)
Comment (dmatveev) [reply modules/gapi/misc/python/python_bridge.hpp:14]: +1

File: modules/gapi/src/backends/common/serialization.cpp (hunk 3 above)
Comment (dmatveev) [reply modules/gapi/src/backends/common/serialization.cpp:369]: +1

File: modules/gapi/src/backends/common/serialization.cpp [hunk 5]
```diff
@@ -366,6 +366,17 @@ IIStream& operator>> (IIStream& is,       cv::gapi::wip::draw::Line &l) {
...
+    std::string statusStr;
+    is >> statusStr;
+    status = cv::gapi::ot::from_string(statusStr);
+    return is;
```
Comment (dmatveev) [inline modules/gapi/src/backends/common/serialization.cpp:373]: Please don't do this. It is binary serialization, no the formatted input/output

File: modules/gapi/src/backends/common/serialization.hpp [hunk 6]
```diff
@@ -130,6 +130,9 @@ GAPI_EXPORTS IIStream& operator>> (IIStream& is,       cv::gapi::wip::draw::Poly
 GAPI_EXPORTS IOStream& operator<< (IOStream& os, const cv::gapi::wip::draw::Line &l);
 GAPI_EXPORTS IIStream& operator>> (IIStream& is,       cv::gapi::wip::draw::Line &l);
 
+GAPI_EXPORTS IOStream& operator<< (IOStream& os, const cv::gapi::ot::TrackingStatus& status);
+GAPI_EXPORTS IIStream& operator>> (IIStream& is,       cv::gapi::ot::TrackingStatus& status);
```
Comment (dmatveev) [inline modules/gapi/src/backends/common/serialization.hpp:134]: Is it required now for status? I believe s11n can be omitted for the tracking status.

File: modules/gapi/test/cpu/gapi_ot_tests_cpu.cpp [hunk 7]
```diff
@@ -223,7 +206,7 @@ TEST(VASObjectTracker, PipelineTest)
             reference_trackings_det_ids[frame_no].push_back(fn_tracked_box["id"]);
             reference_trackings_tr_ids[frame_no].push_back(int(fn_tracked_box["tracking_id"]));
             reference_trackings_tr_statuses[frame_no].push_back(
-                from_string(fn_tracked_box["tracking_status"]));
+                cv::gapi::ot::from_string(fn_tracked_box["tracking_status"]));
```
Comment (dmatveev) [inline modules/gapi/test/cpu/gapi_ot_tests_cpu.cpp:209]: it shouldn't be an API function.

File: modules/python/src2/cv2_convert.hpp [hunk 8]
```diff
@@ -156,6 +156,33 @@ struct PyOpenCV_Converter
...
+            return true;
+        if(PyInt_Check(obj))
+            value = (uint64_t)PyInt_AsUnsignedLongLongMask(obj);
+        else if(PyLong_Check(obj))
+            value = (uint64_t)PyLong_AsUnsignedLongLong(obj);
+        else
+            return false;
+        return value != (uint64_t)-1 || !PyErr_Occurred();
+    }
+};
```
Comment (dmatveev) [inline modules/python/src2/cv2_convert.hpp:183]: Are there tests on that?

File: modules/gapi/misc/python/python_bridge.hpp (hunk 1 above)
Comment (AsyaPronina) [reply modules/gapi/misc/python/python_bridge.hpp]: It couldn't be that time, as was included in `gcommon.hpp`. Now, it is removed from `gcommon,hpp` and returned back to `ot.hpp`

**Comment by AsyaPronina** [comment]
> The changes looks good but invasive. I'm not sure that `OpaqueKind` was ever designed to contain such specific detail as `TrackingStatus`.
> 
> Perhaps we can avoid exposing `TrackingStatus` into python and just use `int` instead. For improving readability we still can define `TrackingStatus` enum but on python side (`__init__.py`) and use it like that:
> 
> ```
> rois, labels, ids, status = cv.gapi.ot.track(img, rect, labels, delta)
> if status == TrackingStatus.NEW.value:
>     pass
> ```
> 
> In that case the `status` is `int` but you may compare it with `enum`:
> 
> ```
> class TrackingStatus(Enum):
>     NEW = 1,
>     TRACKED = 2,
>     LOST = 3
> ```

Fixed!

File: modules/gapi/src/backends/common/serialization.cpp (hunk 3 above)
Comment (AsyaPronina) [reply modules/gapi/src/backends/common/serialization.cpp]: Removed!

**Comment by AsyaPronina** [comment]
> @AsyaPronina Could you add a couple of python test cases, please?

Fixed!

File: modules/gapi/include/opencv2/gapi/gcommon.hpp (hunk 4 above)
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/gcommon.hpp]: Fixed!

File: modules/gapi/src/backends/common/serialization.cpp (hunk 5 above)
Comment (AsyaPronina) [reply modules/gapi/src/backends/common/serialization.cpp]: Removed!

File: modules/gapi/src/backends/common/serialization.hpp (hunk 6 above)
Comment (AsyaPronina) [reply modules/gapi/src/backends/common/serialization.hpp]: Fixed!

File: modules/gapi/test/cpu/gapi_ot_tests_cpu.cpp (hunk 7 above)
Comment (AsyaPronina) [reply modules/gapi/test/cpu/gapi_ot_tests_cpu.cpp]: Fixed!

File: modules/python/src2/cv2_convert.hpp (hunk 8 above)
Comment (AsyaPronina) [reply modules/python/src2/cv2_convert.hpp:183]: Added!

File: modules/gapi/include/opencv2/gapi/ot.hpp [hunk 9]
```diff
@@ -24,7 +24,7 @@ namespace ot {
  *
  * Tracking status twin for vas::ot::TrackingStatus
  */
-enum class TrackingStatus: int32_t
```
[the remark this answers is not in the document]
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/ot.hpp:27]: to be automatically convertable to int

File: modules/gapi/include/opencv2/gapi/ot.hpp [hunk 10]
```diff
@@ -107,23 +107,25 @@ G_API_OP(GTrackFromFrame, <GTrackedInfo(cv::GFrame, cv::GArray<cv::Rect>, cv::GA
...
-                        cv::GArray<int32_t>,
-                        cv::GArray<uint64_t>,
-                        cv::GArray<TrackingStatus>> track(const cv::GMat& mat,
-                                                          const cv::GArray<cv::Rect>& detected_rects,
-                                                          const cv::GArray<int>& detected_class_labels,
-                                                          float delta);
+GAPI_EXPORTS_W std::tuple<cv::GArray<cv::Rect>,
+                          cv::GArray<int>,
+                          cv::GArray<uint64_t>,
+                          cv::GArray<int>>      track(const cv::GMat& mat,
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/ot.hpp:121]: minor: please put `track()` on a new line after its monstrous return value type

File: modules/gapi/misc/python/package/gapi/__init__.py [hunk 11]
```diff
@@ -56,6 +56,14 @@ class Int():
         def __new__(self):
             return cv.GOpaqueT(cv.gapi.CV_INT)
 
+    class Int64():
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/misc/python/package/gapi/__init__.py:59]: One of the return types of OT I guess.

File: modules/gapi/include/opencv2/gapi/ot.hpp [hunk 12]
```diff
@@ -139,15 +141,15 @@ GAPI_EXPORTS std::tuple<cv::GArray<cv::Rect>,
...
-GAPI_EXPORTS std::tuple<cv::GArray<cv::Rect>,
-                        cv::GArray<int32_t>,
-                        cv::GArray<uint64_t>,
-                        cv::GArray<TrackingStatus>> track(const cv::GFrame& frame,
-                                                          const cv::GArray<cv::Rect>& detected_rects,
-                                                          const cv::GArray<int>& detected_class_labels,
-                                                          float delta);
+GAPI_EXPORTS_W std::tuple<cv::GArray<cv::Rect>,
+                         cv::GArray<int>,
+                         cv::GArray<uint64_t>,
```
[the remark this answers is not in the document]
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/ot.hpp:148]: This is what inner VAS function operates with.  I don't think that `reinterpret_cast` or copying is the best option to deal with it and return another type.

**Review by dmatveev — APPROVED** [review]
👍

File: modules/gapi/misc/python/package/gapi/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (AsyaPronina) [reply modules/gapi/misc/python/package/gapi/__init__.py:59]: `UInt64` will be used if we want to create custom kernel in python and accept output from `track` operation.

File: modules/gapi/misc/python/package/gapi/__init__.py [hunk 13]
```diff
@@ -205,6 +225,8 @@ def op(op_id, in_types, out_types):
     type2str = {
         cv.gapi.CV_BOOL:      'cv.gapi.CV_BOOL' ,
         cv.gapi.CV_INT:       'cv.gapi.CV_INT' ,
+        cv.gapi.CV_INT64:     'cv.gapi.CV_INT64' ,
```
[the remark this answers is not in the document]
Comment (AsyaPronina) [reply modules/gapi/misc/python/package/gapi/__init__.py:228]: It was exposed in the bindings already. I decided to just add it here also.

File: modules/gapi/misc/python/test/test_gapi_ot.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    message = str(e)
+
+    class TestSkip(unittest.TestCase):
+        def setUp(self):
+            self.skipTest('Skip tests: ' + message)
+
+        def test_skip():
+            pass
+
+    pass
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/misc/python/test/test_gapi_ot.py:54]: +1

File: modules/gapi/misc/python/test/test_gapi_ot.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+            g_out_rects, g_out_rects_cls, g_track_ids, g_track_sts = \
+                cv.gapi.ot.track(g_in, g_in_rects, g_in_rects_cls, delta)
+
+
+            comp = cv.GComputation(cv.GIn(g_in, g_in_rects, g_in_rects_cls),
+                                   cv.GOut(g_out_rects, g_out_rects_cls,
+                                           g_track_ids, g_track_sts))
+
+            res = comp.apply(cv.gin(in_image, in_rects, in_rects_cls),
+                             args=cv.gapi.compile_args(cv.gapi.ot.cpu.kernels()))
```
Comment (dmatveev) [inline modules/gapi/misc/python/test/test_gapi_ot.py:41]: Should the output be tested somehow? How the LOST/NEW/TRACK status look in Python now?

**Review by TolyaTalamanov — APPROVED** [review]

File: modules/gapi/misc/python/test/test_gapi_ot.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (AsyaPronina) [reply modules/gapi/misc/python/test/test_gapi_ot.py]: Fixed

File: modules/gapi/include/opencv2/gapi/ot.hpp (hunk 10 above)
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/ot.hpp]: Fixed

File: modules/gapi/misc/python/test/test_gapi_ot.py (hunk 15 above)
Comment (AsyaPronina) [reply modules/gapi/misc/python/test/test_gapi_ot.py]: Fixed!

**Comment by AsyaPronina** [comment]
Hello Dear Alexander (@asmorkalov), Dear Alexander (@opencv-alalek), Dear Maksim (@mshabunin), could you please merge this PR?

**Comment by smirnov-alexey** [comment]
@asmorkalov @mshabunin can we merge it today pretty please?

**Review by mshabunin — APPROVED** [review]
Looks good to me.

**Pull request merged**